packages/coordinator/ts/deployer/deployer.service.ts

```diff
@@ -0,0 +1,943 @@
+                  tallyZkeys,
+                ],
+              }),
+            }),
+          },
+        });
+
+        const receiptVkRegistry = await bundlerClient.waitForUserOperationReceipt({
+          hash: userOpHashVkRegistry,
+        });
+
+        if (!receiptVkRegistry.success) {
+          throw new Error(ErrorCodes.FAILED_TO_SET_VERIFYING_KEYS_ON_VK_REGISTRY);
+        }
+      }
+    } catch (error) {
+      this.logger.error("Failed to set verifying keys on vk registry: ", error);
+    }
+
+    // 5. maci (here we don't check whether one is already deployed, we just deploy it)
+    const emptyBallotRoots = genEmptyBallotRoots(config.MACI.stateTreeDepth);
+
+    const maciAddress = await this.deployAndGetAddress(
+      kernelClient,
+      MACIFactory.abi,
+      MACIFactory.linkBytecode({
+        "contracts/crypto/PoseidonT3.sol:PoseidonT3": poseidonT3Address,
+        "contracts/crypto/PoseidonT4.sol:PoseidonT4": poseidonT4Address,
+        "contracts/crypto/PoseidonT5.sol:PoseidonT5": poseidonT5Address,
+        "contracts/crypto/PoseidonT6.sol:PoseidonT6": poseidonT6Address,
+      }) as Hex,
+      [
+        pollFactoryAddress,
+        messageProcessorFactoryAddress,
+        tallyFactoryAddress,
+        gatekeeperAddress,
+        initialVoiceCreditProxyAddress,
+        config.MACI.stateTreeDepth,
+        emptyBallotRoots,
+      ],
+      bundlerClient,
+      publicClient,
+    );
+
+    if (!maciAddress) {
+      throw new Error(`${ErrorCodes.FAILED_TO_DEPLOY_CONTRACT} - ${EContracts.MACI}`);
+    }
+
+    await this.storage.register({
+      id: EContracts.MACI,
+      contract: new BaseContract(maciAddress, MACIFactory.abi as unknown as InterfaceAbi),
+      args: [
+        pollFactoryAddress,
+        messageProcessorFactoryAddress,
+        tallyFactoryAddress,
+        gatekeeperAddress,
+        initialVoiceCreditProxyAddress,
+        config.MACI.stateTreeDepth,
+        emptyBallotRoots.map((root) => root.toString()),
+      ],
+      network: chain,
+    });
+
+    // set the gate on the gatekeeper
+    const userOpHash = await kernelClient.sendUserOperation({
+      userOperation: {
+        callData: await kernelClient.account.encodeCallData({
+          to: gatekeeperAddress! as Hex,
+          value: 0n,
+          data: encodeFunctionData({
+            abi: gatekeeperData.abi,
+            functionName: "setMaciInstance",
+            args: [maciAddress],
+          }),
+        }),
+      },
+      account: kernelClient.account,
+    });
+
+    const receipt = await bundlerClient.waitForUserOperationReceipt({
+      hash: userOpHash,
+    });
+
+    if (!receipt.success) {
+      throw new Error(ErrorCodes.FAILED_TO_SET_MACI_INSTANCE_ON_GATEKEEPER);
+    }
+
+    return maciAddress;
+  }
+
+  /**
+   * Deploy a poll
+   *
+   * @param args - deploy poll dto
+   * @returns poll id
+   */
+  async deployPoll({ approval, sessionKeyAddress, chain, config }: IDeployPollArgs): Promise<string> {
```

Comment:
would the registry contract part updated in another PR?

Reply:
oh right yea let's do that separately once this is merged to not make it even bigger